go/arrow/memory/mallocator/mallocator.go

```diff
@@ -0,0 +1,97 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build cgo
```

Comment:
nothing wrong with this, just noting that technically it's unneccesary as Go will automagically exclude any file that contains `import "C"` if cgo is disabled. 

However, we may want to put this behind a feature tag potentially that also changes the default allocator to use this? So that someone can easily just add `tags malloc` or whatever to their build line and the default allocator uses malloc instead of the Go one?

Reply:
I gave it a shot, but it creates an import cycle between `mallocator` and `memory`.

Reply:
The only thing you're using `memory` for is the definition of that `TestingT` interface right? Just make a local interface with the same definition and you don't need to import `memory` and you don't lose any functionality.

Reply:
Ah right, structural typing...

Reply:
Done! Also added a docstring for mallocator